### ISSUE TITLE
XmlException: Text node cannot appear in this state

### ISSUE THREAD

**Issue by DropboxKenshiro** [BayatGames/Support#2]
Firstly, thank you for great addon for saving games! It really helped me!
Hovever, I came across some weird errors while using XML serializer.

So, what I want to do is save state of building in my game to hold it between sessions.  Following the way shown in documentation, I created a class to hold data. All objects operate on single object of that class. Then I use loop to iterate trough all instances of buildings.

And here errors appear! Here's some of them:
"XmlException: Text node cannot appear in this state.  Line 13, position 9.
Mono.Xml2.XmlTextReader.ReadText (Boolean notWhitespace)
Mono.Xml2.XmlTextReader.ReadContent ()
Mono.Xml2.XmlTextReader.Read () etc."
and
"XmlException: Text node cannot appear in this state.  Line 13, position 9.
_(this supposed to be sth with null in Unity console)_
Mono.Xml2.XmlTextReader.ReadText (Boolean notWhitespace)
Mono.Xml2.XmlTextReader.ReadContent ()
Mono.Xml2.XmlTextReader.Read ()
System.Xml.XmlTextReader.Read ()
System.Xml.XmlReader.ReadEndElement ()
System.Xml.Serialization.XmlSerializationReader.ReadEndElement ()
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstance (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObject (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot (System.Xml.Serialization.XmlTypeMapping rootMap)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot ()
System.Xml.Serialization.XmlSerializer.Deserialize (System.Xml.Serialization.XmlSerializationReader reader)
UnityEngine.Debug:LogException(Exception)
SaveGameFree.Serializers.SaveGameFreeXmlSerializer:Deserialize(String) (at Assets/SaveGameFree/Serializers/SaveGameFreeXmlSerializer.cs:50)
SaveGameFree.Saver:Load(String, ISerializer, PathType, String) (at Assets/SaveGameFree/Saver.cs:628)
SaveGameFree.Saver:Load(String) (at Assets/SaveGameFree/Saver.cs:539)
SaveLoad:Awake() (at Assets/Skrypty/Widok miasta/SaveLoad.cs:18)
"

Both are mysterious for me.

So, I looking forward for a reply.
Dropbox

**Comment by hasanbayatme** [comment]
Hi and thanks for your awesomeness.

It is mysterious error.

Can you send your iteration script?

Do you tried to change encoding for the xml serializer?

Thanks.

**Comment by DropboxKenshiro** [comment]
Here's the saving code:
```csharp
void Save()
    {
        OverallParameters[] buildings = GetComponentsInChildren<OverallParameters>(); // it's here to catch all buildings without looping too much
        foreach(OverallParameters p in buildings)
        {
            data.buildings.Add(p.dataoverall); // dataoverall is class for saving
        }

        SaveGameFree.Saver.Save(data, insertnamehere);
        Debug.Log("Saved");
    }
```

But I really have no clue where coding is.

**Comment by hasanbayatme** [comment]
What is your dataoverall class? is it a complex class? if is a complex class, you should implement ISerializable interface to add serialization functionally.

You can only save classes that have Serializable attribute or implement ISerializable interface, otherwise the class won't saved and throws exception.

Here is too late (i am going to sleep), i can't respond to you for a period, but i will respond to you when i wake up.

Thanks.

**Comment by hasanbayatme** [comment]
Do you tried the above suggestions?

Thanks.

**Comment by DropboxKenshiro** [comment]
OverallData class have `System.Serializable` parameter. Hovever, I'm 🔰 in Unity, so I don't even know when class is complex or not and what's an interface. _Of course I could learn now_
I tried to google complex class, but I got only complex numbers.
And, about XML file coding, I just searched for it in scripts, but I can't find it.

EDIT: Oh. The coding could be specified in `Serialize` function. Hovever, overload for that is mysterious for me.

**Comment by hasanbayatme** [comment]
Can you post your OverallData script?

I mean by complex classes is large classes that contains several properties and methods.

Also you can modify the source code of save game free xml serialization at Assets/SaveGameFree/Serializers/SaveGameFreeXmlSerializer.cs 👍 

Interfaces are as their name shown, interfaces are useful utilities that shows the programmer how to use the interface methods and properties, this makes job a lot easier, and is one the basics in programming:

```csharp
public interface ExampleInterface {
  void DoSomeJob ();
}

public class ExampleClass : ExampleInterface {
  public void DoSomeJob () {
    Console.WriteLine ("Hello World");
  }
}
```

Also you can make your own xml serialization for save game free, you can try [YAXlib](https://github.com/sinairv/YAXLib) or [SharpSerializer](https://sharpserializer.codeplex.com/). (this might solve your problem 🥇 )

[Learn more about Save Game Free Serializers](https://github.com/EmpireAssets/SaveGameFree/wiki/How-to-Use-Serializer%3F)

Thanks.

**Comment by DropboxKenshiro** [comment]
Thank you for your help.
Firstly, I implemented ISerializer for saving and used `XMLWriter` for encoding setting. Well, it's not that thing.

So, here's OverallData:
```csharp
[System.Serializable]
public class OverallData
{
   //some Polish here, hovever types are key
    public Vector2 pozycja;
    public Sprite sprajt;

    public string nazwa;

    public bool ukończony = false;
    public int wytrzymałość = 0;

    public int[] surowce_potrzebne = new int[8];
    public int[] surowce_dostarczone = new int[8];

    public int robotnicy = 0;
    public int wymagani = 0;

    public int ad_start;
    public int ad_koniec;

    public int atrakcyjność;

    public ArrayList dodatkowe_dane;   
}
```

Additionally, here's main save class:
```csharp
[System.Serializable]
public class Zapis : SaveGameFree.ISerializer
{
    public int[] materiały = new int[8];

    public List<OgolneDane> budynki = new List<OgolneDane>();
```

Also, I found that second error in my first post was probably a copy of first one, so here's new thing:
```
NullReferenceException
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:222)
Rethrow as TargetInvocationException: Exception has been thrown by the target of an invocation.
System.Reflection.MonoMethod.Invoke (System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:232)
System.Reflection.MonoProperty.SetValue (System.Object obj, System.Object value, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] index, System.Globalization.CultureInfo culture) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoProperty.cs:331)
System.Reflection.PropertyInfo.SetValue (System.Object obj, System.Object value, System.Object[] index) (at /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/PropertyInfo.cs:102)
System.Xml.Serialization.XmlTypeMapMember.SetValue (System.Object ob, System.Object value)
System.Xml.Serialization.XmlSerializationReaderInterpreter.SetMemberValue (System.Xml.Serialization.XmlTypeMapMember member, System.Object ob, System.Object value, Boolean isValueList)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadMembers (System.Xml.Serialization.ClassMap map, System.Object ob, Boolean isValueList, Boolean readByOrder)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstanceMembers (System.Xml.Serialization.XmlTypeMapping typeMap, System.Object ob)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstance (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObject (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObjectElement (System.Xml.Serialization.XmlTypeMapElementInfo elem)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadMembers (System.Xml.Serialization.ClassMap map, System.Object ob, Boolean isValueList, Boolean readByOrder)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstanceMembers (System.Xml.Serialization.XmlTypeMapping typeMap, System.Object ob)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstance (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObject (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObjectElement (System.Xml.Serialization.XmlTypeMapElementInfo elem)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadListElement (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, System.Object list, Boolean canCreateInstance)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadMembers (System.Xml.Serialization.ClassMap map, System.Object ob, Boolean isValueList, Boolean readByOrder)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstanceMembers (System.Xml.Serialization.XmlTypeMapping typeMap, System.Object ob)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadClassInstance (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadObject (System.Xml.Serialization.XmlTypeMapping typeMap, Boolean isNullable, Boolean checkType)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot (System.Xml.Serialization.XmlTypeMapping rootMap)
System.Xml.Serialization.XmlSerializationReaderInterpreter.ReadRoot ()
System.Xml.Serialization.XmlSerializer.Deserialize (System.Xml.Serialization.XmlSerializationReader reader)
UnityEngine.Debug:LogException(Exception)
SaveGameFree.Serializers.SaveGameFreeXmlSerializer:Deserialize(String) (at Assets/SaveGameFree/Serializers/SaveGameFreeXmlSerializer.cs:50)
SaveGameFree.Saver:Load(String, ISerializer, PathType, String) (at Assets/SaveGameFree/Saver.cs:628)
SaveGameFree.Saver:Load(String) (at Assets/SaveGameFree/Saver.cs:539)
SaveLoad:Awake() (at Assets/Skrypty/Widok miasta/SaveLoad.cs:21)
```

Probably I'll try to use another library and try to import it into Unity.

EDIT: Strangly enough, error appears even save haven't been presented yet.

**Comment by hasanbayatme** [comment]
There is no need to use XmlWriter or XmlReader, you can just use the Xml serialization libraries API that i have posted to serialize data.

Some data types aren't serializable, such as `Vector3`, `Sprite`, ... so you should avoid them in your data.

Why you are trying to save Sprite type?

So, if you want to save Vector2 you should create an replacement for Vector2, create a new c# script file and name it Vector2Savve and put the below content in it:

```csharp
using System;
using System.Collections;
using System.Collections.Generic;
using UnityEngine;

namespace SaveGameFree.Types
{
	
	/// <summary>
	/// Representation of 2D vectors and points.
	/// </summary>
	[Serializable]
	public struct Vector2Save
	{
	
		public float x;
		public float y;

		public Vector2Save ( float x )
		{
			this.x = x;
			this.y = 0f;
		}

		public Vector2Save ( float x, float y )
		{
			this.x = x;
			this.y = y;
		}

		public Vector2Save ( Vector2 vector )
		{
			this.x = vector.x;
			this.y = vector.y;
		}

		public Vector2Save ( Vector3 vector )
		{
			this.x = vector.x;
			this.y = vector.y;
		}

		public Vector2Save ( Vector4 vector )
		{
			this.x = vector.x;
			this.y = vector.y;
		}

		public Vector2Save ( Vector2Save vector )
		{
			this.x = vector.x;
			this.y = vector.y;
		}

		public static implicit operator Vector2Save ( Vector2 vector )
		{
			return new Vector2Save ( vector );
		}

		public static implicit operator Vector2 ( Vector2Save vector )
		{
			return new Vector2 ( vector.x, vector.y );
		}

		public static implicit operator Vector2Save ( Vector3 vector )
		{
			return new Vector2Save ( vector );
		}

		public static implicit operator Vector3 ( Vector2Save vector )
		{
			return new Vector3 ( vector.x, vector.y );
		}

		public static implicit operator Vector2Save ( Vector4 vector )
		{
			return new Vector2Save ( vector );
		}

		public static implicit operator Vector4 ( Vector2Save vector )
		{
			return new Vector4 ( vector.x, vector.y );
		}

	}

}
```

And then replace the Vector2 in your OverallData with the new created Vector2Save class, and also comment out the `sprajt` because it is an unserializable type: (It is not recommended to use any language other than English for naming variables and methods)

```csharp
[System.Serializable]
public class OverallData
{
   //some Polish here, hovever types are key
    public Vector2Save pozycja;
    //public Sprite sprajt; we can't save it

    public string nazwa;

    public bool ukończony = false;
    public int wytrzymałość = 0;

    public int[] surowce_potrzebne = new int[8];
    public int[] surowce_dostarczone = new int[8];

    public int robotnicy = 0;
    public int wymagani = 0;

    public int ad_start;
    public int ad_koniec;

    public int atrakcyjność;

    public ArrayList dodatkowe_dane;   
}
```

Then try to use the previous serializer (deafult Save Game Free xml serializer) to save your data and check out the result.

Also what is your **Zapis** class? Why you marked it Serializable while implementing ISerializer?

ISerializer is an interface for Serializers not for Serializables.

Also in your `result = (T)serializer.Deserialize(czytelnik, "utf-8");` the utf-8 is invalid, because the parameter does not accepts encoding, it accepts encoding styles for xml. [See documentation](https://msdn.microsoft.com/en-us/library/dk9cbaf1(v=vs.110).aspx)

[:book: Checkout this tutorial](https://github.com/EmpireAssets/SaveGameFree/wiki/How-to-Create-Custom-Serializer%3F) to learn how to create a serializer.

Thanks.

**Comment by DropboxKenshiro** [comment]
That was A PAIN!
Ok, so, thank you.
My original idea was saving whole object. Vector2 is necessary to save position beteween sessions, to recreate buildings position again. I think I'll just create an array of building prefabs and I just save a index. Thank for Vector replacement idea. I think i'll rework save system.
Also small proposition: maybe you could add list of non-serializable types.
Really thank you for all help!

**Comment by DropboxKenshiro** [comment]
After removing non serializables, save works without errors.
Solved.
I'm closing.

**Comment by hasanbayatme** [comment]
I'm so happy to have solved your problem.

It is not pain, just replace your Vector2 where you want to save, and all of your Vector2 will be automatically converted to Vector2Save.

What save system you want to implement?

I will submit a list of them to Wiki when i released Save Game Free 2.0.

Let me know if you need help.

Thanks.

EDIT: Excuse me for my late responses, i am working on Save Game Free 2.0
